### PR TITLE
Update Neo4j Integration to Support Neo4j 4.x

### PR DESCRIPTION
**Pull request by davidfauth**
### What does this PR do?

Rewrote the Neo4j integration to support Neo4j 4.x. 
Used Neo4j integration that is currently used in our Aura product.

### Motivation

Update integration for 4.x to support customers using Neo4j 4.x

### Review checklist

- [ ] PR has a [meaningful title](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title) or PR has the `no-changelog` label attached
- [ ] Feature or bugfix has tests
- [ ] Git history is clean
- [ ] If PR impacts documentation, docs team has been notified or an issue has been opened on the [documentation repo](https://github.com/DataDog/documentation/issues/new)

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: neo4j/README.md
Comment (FlorianVeaux) [inline neo4j/README.md:15]: You should leave this current section, Datadog users will be able to install your integration automatically with `datadog-agent integration install -t datadog-neo4j==4.0.0` once this PR is merged

File: neo4j/README.md
Comment (FlorianVeaux) [inline neo4j/README.md:48]: Need to fill this

File: neo4j/assets/service_checks.json
Comment (FlorianVeaux) [inline neo4j/assets/service_checks.json:1]: Why are you removing this file?

File: neo4j/datadog_checks/neo4j/data/conf.yaml.example
Comment (FlorianVeaux) [inline neo4j/datadog_checks/neo4j/data/conf.yaml.example:12]: You're going to break backwards compatibility for users already running the neo4j integration. Nothing blocking, just want to make sure you know about this

File: neo4j/datadog_checks/neo4j/neo4j.py
Comment (FlorianVeaux) [inline neo4j/datadog_checks/neo4j/neo4j.py:34]: FYI: You don't provide the ability to use https here.

File: neo4j/datadog_checks/neo4j/neo4j.py
Comment (FlorianVeaux) [inline neo4j/datadog_checks/neo4j/neo4j.py:28]: Two optional things here:
1. You should probably extend from `OpenMetricsBaseCheck` instead, it's a better option and is now the standard for all our official integrations. It provides more configuration options, and it's more maintained.
2. Because you extend from PrometheusCheck (or `OpenMetricsBaseCheck` if you change it), you have much more configuration option available to you. As an example you can see how we do it for a simple integration here: https://github.com/Datadog/integrations-core/blob/master/scylla/datadog_checks/scylla/scylla.py#L32, for generating the [conf.yaml.example](https://github.com/DataDog/integrations-core/blob/master/scylla/datadog_checks/scylla/data/conf.yaml.example) file automatically with all options we use this [spec file](https://github.com/DataDog/integrations-core/blob/master/scylla/assets/configuration/spec.yaml). If you want your users to benefit from all the config options you can do something similar

File: neo4j/tox.ini
Comment (FlorianVeaux) [inline neo4j/tox.ini:4]:
```suggestion
basepython = py38
```

File: neo4j/manifest.json
Comment (FlorianVeaux) [inline neo4j/manifest.json:17]:
```suggestion
  "public_title": "Datadog-Neo4j Integration",
```

**Comment by platinummonkey** [comment]
not sure why I'm tagged here :) dismissing my review

File: neo4j/README.md
Comment (kayayarai) [inline neo4j/README.md:17]: Throughout: Links need to be changed to the style `[linktext][x]` with URLs collected at the bottom of the file.

File: neo4j/datadog_checks/neo4j/data/conf.yaml.example
Comment (hithwen) [reply neo4j/datadog_checks/neo4j/data/conf.yaml.example:12]: You can avoid this by reading both options in the code

File: neo4j/README.md
Comment (davidfauth) [reply neo4j/README.md]: Updated section with latest commit.

File: neo4j/datadog_checks/neo4j/data/conf.yaml.example
Comment (davidfauth) [reply neo4j/datadog_checks/neo4j/data/conf.yaml.example:12]: We made a lot of changes in 4.x so wanted a clean integration.
This code base will work with 3.5.

File: neo4j/tox.ini
Comment (davidfauth) [reply neo4j/tox.ini]: I believe I have changed this.

File: neo4j/manifest.json
Comment (davidfauth) [reply neo4j/manifest.json]: I believe this has been changed as well.

File: neo4j/assets/service_checks.json
Comment (davidfauth) [reply neo4j/assets/service_checks.json]: Newest version has no service checks. Is that a requirement?

File: neo4j/datadog_checks/neo4j/neo4j.py
Comment (davidfauth) [reply neo4j/datadog_checks/neo4j/neo4j.py]: Added in ability to use https.

File: neo4j/datadog_checks/neo4j/neo4j.py
Comment (davidfauth) [reply neo4j/datadog_checks/neo4j/neo4j.py:30]: Will need to look at this in a future release.

File: neo4j/README.md
Comment (davidfauth) [reply neo4j/README.md:17]: I think I have these all changed.

File: neo4j/assets/service_checks.json
Comment (yzhan289) [inline neo4j/assets/service_checks.json:17]: Why delete the `can_connect` service check?

File: neo4j/datadog_checks/neo4j/neo4j.py
Comment (yzhan289) [inline neo4j/datadog_checks/neo4j/neo4j.py:5]: This used to be an `AgentCheck`, but you change it to `PrometheusCheck`. There may be some users using the older versions of this check, but up to you if you want to get rid of the old logic completely.

**Review by yzhan289 — COMMENTED** [review]
Mostly fine, just some confusion related to backwards compatibility. Also looks like there are some merge conflicts.

File: neo4j/datadog_checks/neo4j/neo4j.py
Comment (davidfauth) [reply neo4j/datadog_checks/neo4j/neo4j.py:5]: Going to leave it as is.

File: neo4j/datadog_checks/neo4j/data/conf.yaml.example
Comment (yzhan289) [inline neo4j/datadog_checks/neo4j/data/conf.yaml.example:50]: Seems like CI is failing because it doesn't like use of `[1]` and `[2]` for links. I think keeping it the old way is fine.

**Review by yzhan289 — CHANGES_REQUESTED** [review]

File: neo4j/datadog_checks/neo4j/data/conf.yaml.example
Comment (davidfauth) [reply neo4j/datadog_checks/neo4j/data/conf.yaml.example:50]: Went back to old format. Let me know if that works.

Comment (yzhan289) [reply neo4j/datadog_checks/neo4j/data/conf.yaml.example:50]: I don't think you pushed your changes yet.

File: neo4j/README.md
Comment (kayayarai) [reply neo4j/README.md:17]: Yes, except for the former link [1], in step 1 which Florian suggests reinstating (replacing your current steps 1-3, I think).

File: neo4j/README.md
Comment (kayayarai) [inline neo4j/README.md:19]:
```suggestion
4. To install the neo4j check on your host:
```

**Review by kayayarai — CHANGES_REQUESTED** [review]

File: neo4j/assets/configuration/spec.yaml
Comment (hithwen) [inline neo4j/assets/configuration/spec.yaml:1]: Why remove the configuration?

File: neo4j/assets/service_checks.json
Comment (hithwen) [inline neo4j/assets/service_checks.json:1]: why remove the service checks?

File: neo4j/datadog_checks/neo4j/__about__.py
Comment (hithwen) [inline neo4j/datadog_checks/neo4j/__about__.py:2]: This is the version of the check not the version of neo4j, probably you can make it 1.2.0 or 2.0.0

File: neo4j/README.md
Comment (hithwen) [inline neo4j/README.md:17]:
```suggestion
```
This is not required

File: neo4j/README.md
Comment (hithwen) [inline neo4j/README.md:18]:
```suggestion
3. Download and install the [Datadog Agent](https://app.datadoghq.com/account/settings#agent).
```

File: neo4j/bin/generate-metadata-content
Comment (hithwen) [inline neo4j/bin/generate-metadata-content:3]: what's this script for?

File: neo4j/datadog_checks/neo4j/neo4j.py
Comment (hithwen) [inline neo4j/datadog_checks/neo4j/neo4j.py:5]: you have removed the actual check?

Comment (hithwen) [inline neo4j/datadog_checks/neo4j/neo4j.py:5]: which class inherits from AgentCheck now?

File: neo4j/metadata.csv
Comment (hithwen) [inline neo4j/metadata.csv:2]: You should not remove old metrics since there can be users running older versions of the check. What you can do is mention in the description that the metrics is only emitted up to version X of the check

File: neo4j/datadog_checks/neo4j/data/conf.yaml.example
Comment (FlorianVeaux) [inline neo4j/datadog_checks/neo4j/data/conf.yaml.example:70]:
```suggestion
```
This is invalid yaml, you'll need to remove these two lines

File: neo4j/datadog_checks/neo4j/data/conf.yaml.example
Comment (FlorianVeaux) [inline neo4j/datadog_checks/neo4j/data/conf.yaml.example:56]:
```suggestion
    ## @param min_collection_interval - integer - optional - default: 15
```

File: neo4j/datadog_checks/neo4j/data/conf.yaml.example
Comment (FlorianVeaux) [inline neo4j/datadog_checks/neo4j/data/conf.yaml.example:12]:
```suggestion
  - # host: localhost
```

**Comment by FlorianVeaux** [comment]
Thanks for the few changes you've made, FYI the previous comments still need addressing.

**Comment by ashley-abbott** [comment]
@davidfauth have you had any chance to look at this, I could really do with v4.x being supported

**Comment by davidfauth** [comment]
@ashley-abbott addressed some additional issues just now

**Comment by ashley-abbott** [comment]
> @ashley-abbott addressed some additional issues just now

thanks for the swift action @davidfauth

**Comment by davidfauth** [comment]
@ashley-abbott is there any documentation on when / how to use histogram properly within the metadata.csv?

**Review by ofek — CHANGES_REQUESTED** [review]
Since we're introducing breaking changes, WDYTA dropping support for Python 2 and using the latest `OpenMetricsBaseCheckV2`? Here are some examples, with increasing levels of complexity:

- https://github.com/DataDog/integrations-core/blob/master/kong/datadog_checks/kong/check.py
- https://github.com/DataDog/integrations-core/blob/master/linkerd/datadog_checks/linkerd/check.py
- https://github.com/DataDog/integrations-core/blob/master/avi_vantage/datadog_checks/avi_vantage/check.py
- https://github.com/DataDog/integrations-core/blob/master/amazon_msk/datadog_checks/amazon_msk/check.py

You'll likely only need to define a `get_default_config` method.

Here are all the options: https://github.com/DataDog/integrations-core/blob/master/datadog_checks_dev/datadog_checks/dev/tooling/templates/configuration/instances/openmetrics.yaml

Here's an example of defining the config: https://github.com/DataDog/integrations-core/blob/master/avi_vantage/assets/configuration/spec.yaml

**Comment by yzhan289** [comment]
There's a merge conflict for `neo4j/datadog_checks/neo4j/data/conf.yaml.example`. You can do `ddev validate config neo4j -s` to fix it. Also what @ofek suggested is a good idea. Using `OpenMetricsBaseCheckV2` will allow better performance and different metrics, so if there's a good time to implement a check with that it would be now.  More info on https://github.com/DataDog/integrations-core/pull/8300.

**Comment by davidfauth** [comment]
Fixed code to pass tests locally. Pushed latest update.

File: neo4j/CHANGELOG.md
Comment (ChristineTChen) [reply neo4j/CHANGELOG.md:5]: We should not be trying to match the integration version with the product version.

The accurate next version should be 2.0.0 since we are adding feature support. We shouldn't be removing the previous changelog line 1.0.1, or else we lose historical context.

The changelog should look like:

```
## 2.0.0 / 2021-07-21
* [Added] Updated Neo4j integration to support 4.x. See [#944](https://github.com/DataDog/integrations-extras/pull/944)

## 1.0.1 / 2021-07-28

* [Added] Update neo4j check to use requests wrapper. See [#894](https://github.com/DataDog/integrations-extras/pull/894).
```

We should maybe detail some more of the breaking changes introduced in this PR in the changelog for reference.

File: neo4j/assets/service_checks.json
Comment (ChristineTChen) [reply neo4j/assets/service_checks.json:1]: bump, if you run this version of the integration `datadog-agent check neo4j` you will see a service check is submitted that looks like `neo4j.prometheus.health`. This is submitted in the base check you're inheriting: https://github.com/DataDog/integrations-core/blob/master/datadog_checks_base/datadog_checks/base/checks/prometheus/mixins.py#L595

We should document that in service_checks.json


See docs for more info https://docs.datadoghq.com/developers/service_checks/

File: neo4j/assets/configuration/spec.yaml
Comment (ChristineTChen) [inline neo4j/assets/configuration/spec.yaml:11]: Removing `neo4j_url` is a significant breaking change. Any user who upgrades will have errors and difficulty debugging. The change should be documented in detail in the README.md at minimum (specifically in Configuration section)

File: neo4j/assets/configuration/spec.yaml
Comment (ChristineTChen) [inline neo4j/assets/configuration/spec.yaml:78]: You can utilize configuration templates to provide users will all the configuration options available.

```suggestion
  - template: instance/openmetrics_legacy
```

Then run `ddev validate config -s`.

File: neo4j/assets/configuration/spec.yaml
Comment (ChristineTChen) [inline neo4j/assets/configuration/spec.yaml:82]:
```suggestion

```
This section can also be removed, i couldn't highlight everything in the above suggestion dur to L27 😅

File: neo4j/assets/configuration/spec.yaml
Comment (ChristineTChen) [inline neo4j/assets/configuration/spec.yaml:8]:
```suggestion
          options:
          - template: init_config/openmetrics_legacy
```

File: neo4j/manifest.json
Comment (ChristineTChen) [inline neo4j/manifest.json:31]: this file was deleted, we should replace. This is required

File: neo4j/metadata.csv
Comment (ChristineTChen) [reply neo4j/metadata.csv:2]: bump

**Review by ChristineTChen — CHANGES_REQUESTED** [review]
Hi @davidfauth ,

I've left some suggestions to the changelog, spec.yaml/conf.yaml.example (there's a merge conflict because it's out of sync), and some concerns about backwards incompatibility.

File: neo4j/README.md
Comment (kayayarai) [inline neo4j/README.md:17]:
```suggestion
1. Download and install the [Datadog Agent][8].
```

File: neo4j/README.md
Comment (kayayarai) [inline neo4j/README.md:60]:
```suggestion
[7]: https://docs.datadoghq.com/help
[8]: https://app.datadoghq.com/account/settings#agent
```

**Comment by ofek** [comment]
@davidfauth Hello, this looks nearly done! Could you address the remaining comments?

File: neo4j/README.md
Comment (kayayarai) [inline neo4j/README.md:6]:
```suggestion
Verify the metrics and checks that are submitted through this Agent. Because Neo4j 4.0 and onward supports multiple databases, some metrics and checks are no longer published.
```
What does it mean to "verify" the metrics and checks? Why are we telling users to do this? I can't tell what it means, just that it's a very unusual statement in documentation.

File: neo4j/README.md
Comment (kayayarai) [inline neo4j/README.md:43]:
```suggestion
Service check `neo4j.prometheus.health` is submitted in the base check.
```

**Review by kayayarai — CHANGES_REQUESTED** [review]

File: neo4j/tests/test_neo4j.py
Comment (coignetp) [inline neo4j/tests/test_neo4j.py:33]:
```suggestion
```

**Review by coignetp — CHANGES_REQUESTED** [review]
Could you rebase on latest master or merge latest master in your branch? it shouldn't need to update other integrations config

**Comment by davidfauth** [comment]
Can I get 30 minutes with someone to rebase properly and try to close this out?

**Comment by fanny-jiang** [comment]
Hi @davidfauth, we have Community office hours scheduled this Wednesday if you're able to attend: https://docs.datadoghq.com/developers/community/office_hours/

Please let me know if that works for you, otherwise we can schedule a sync.

**Comment by yzhan289** [comment]
Hi @davidfauth , you can undo the change by running `git revert 30780ff1ec41f23e5f14312827142861d19aa4e9` since it looks like that specific commit was the one that brought in the changes from other integrations. I ran it on my own version of your PR and[ it worked for me](https://github.com/DataDog/integrations-extras/pull/1133).

Once that's done, you can address the other remaining comments and ping me when it's ready for re-review 😄

**Comment by ofek** [comment]
@davidfauth Updated!

1. Agent 7.34 is required (releasing in a week or 2)
2. `host`/`port`/`https` is now just `openmetrics_endpoint`
3. `neo4j_version` is now optional

LMK what you think

**Comment by davidfauth** [comment]
@ofek Looks good to me.

**Review by fanny-jiang — APPROVED** [review]

**Pull request merged**
